File: Scala/src/main/scala/iterator/DinerMenuIterator.scala
package iterator

class DinerMenuIterator(menu: Array[MenuItem]) extends Iterator {
  private var position = 0
  override def hasNext: Boolean = {
    if (position >= menu.length || menu(position) == null) {
      false
    } else {
      true
    }
  }

  override def next: MenuItem = {
    val menuItem = menu(position)
    position += 1
    menuItem
  }
}

File: Kotlin/src/main/kotlin/chapter9_iterator/DinerMenuIterator.kt
package chapter9_iterator

class DinerMenuIterator(
    val menuItems: Array<MenuItem?>,
): Iterator {
    private var pos = 0

    override fun hasNext(): Boolean {
        return !(pos >= menuItems.size || menuItems[pos] == null)
    }

    override fun next(): MenuItem {
        return menuItems[pos++] ?: throw NullPointerException("더이상 출력할 메뉴가 없습니다")
    }
}
